network-mux/src/Network/Mux/Bearer/Socket.hs

```diff
@@ -44,34 +44,59 @@ import qualified Network.Mux.Time as Mx
 -- 'MuxError'.
 --
 socketAsMuxBearer
-  :: Tracer IO Mx.MuxTrace
+  :: Maybe DiffTime
+  -> Tracer IO Mx.MuxTrace
   -> Socket.Socket
   -> MuxBearer IO
-socketAsMuxBearer tracer sd =
+socketAsMuxBearer sduTimeout_m tracer sd =
       Mx.MuxBearer {
         Mx.read    = readSocket,
         Mx.write   = writeSocket,
         Mx.sduSize = 12288
       }
     where
+      hdrLenght = 8
+
       readSocket :: HasCallStack => IO (Mx.MuxSDU, Time)
       readSocket = do
           traceWith tracer $ Mx.MuxTraceRecvHeaderStart
-          hbuf <- recvLen' True 8 []
+
+          -- Wait for the first part of the header without any timeout
+          h0 <- recvAtMost False hdrLenght
+
+          -- Optionally wait at most sduTimeout seconds for the complete SDU.
+          r_m <- case sduTimeout_m of
+                      Just sduTimeout -> timeout sduTimeout $ recvRem h0
+                      Nothing -> Just <$> recvRem h0
+          case r_m of
+                Nothing -> do
+                    traceWith tracer $ Mx.MuxTraceSDUReadTimeout
```

Comment:
A suggestion: maybe let's call this trace `MuxTraceSDUReadTimeoutException` - to make it clearer that we throw an exception.